### PR TITLE
add how services work to how swarm mode works guide

### PR DESCRIPTION
**Pull request by sfsmithcha**
Adds how services work article to the how swarm mode works guide

Signed-off-by: Charles Smith charles.smith@docker.com

### CODE REVIEW

**Comment by sfsmithcha** [comment]
ping @mgoelzer

**Comment by AkihiroSuda** [comment]
typo in service-lifecycle.png: orhcestrator

File: docs/swarm/how-swarm-mode-works/services.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+A task is a one-directional mechanism. It progresses monotonically through a
+series of states: assigned, prepared, running, etc.  If the task fails the
+scheduler removes the task and its container and then creates a new task to
+replace it according to the desired state specified by the service.
+
+The underlying logic of Docker swarm mode is a general purpose scheduler and
+orchestrator.  The service and task abstractions themselves are unaware of the
+containers they implement.  Hypothetically, you could implement other types of
+tasks virtual machine tasks or non-containerized process tasks.  The scheduler
```
Comment (AkihiroSuda) [inline docs/swarm/how-swarm-mode-works/services.md]: "... other types of tasks _such as_ virtual machine tasks..."?

**Comment by AkihiroSuda** [comment]
mouse cursor seems captured in images :mouse:

File: docs/swarm/how-swarm-mode-works/services.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+When you create a service, you specify which container image to use and which
+commands to execute inside running containers. You also define options for the
+service including:
+
+* the port where the swarm will make the service available outside the swarm
+* an overlay network for the service to connect to other services in the swarm
+* CPU and memory limits and reservations
+* a rolling update policy
+*the number of replicas of the image to run in the swarm
```
Comment (SvenDowideit) [inline docs/swarm/how-swarm-mode-works/services.md]: missin a space before the star :)

File: docs/swarm/how-swarm-mode-works/services.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+A container is an isolated process.  In the swarm mode model, each task invokes
+exactly one container. A task is analogous to a “slot” where the scheduler
+places a container. Once the container is live, the scheduler recognizes that
+the task is in a running state.  If the container fails health checks or
+terminates, the task terminates.
+
+## Tasks and scheduling
+
+A task is the atomic unit of scheduling within a swarm.  WHen you declare a
```
Comment (SvenDowideit) [inline docs/swarm/how-swarm-mode-works/services.md]: `s/WHen/When/`

**Comment by mgoelzer** [comment]
LGTM / IANAM

Love the diagrams you did, big improvement over what we had initially!

**Comment by aaronlehmann** [comment]
Can you please avoid spaces in the filenames of the .svg files? It's not a big deal, but it will be a little friendlier for people who encounter these files using a command line interface.

File: docs/swarm/how-swarm-mode-works/services.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+parent="how-swarm-works"
+weight="4"
++++
+<![end-metadata]-->
+
+# How services work
+
+To deploy an application image when Docker Engine is in swarm mode, you create a
+service. Frequently a service will be the image for a microservice within the
+context of some larger application.  Examples of services might include an HTTP
```
Comment (aaronlehmann) [inline docs/swarm/how-swarm-mode-works/services.md]: Nit: extra space

File: docs/swarm/how-swarm-mode-works/services.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+terminates, the task terminates.
+
+## Tasks and scheduling
+
+A task is the atomic unit of scheduling within a swarm.  When you declare a
+desired service state by creating or updating a service, the orchestrator
+realizes the desired state by scheduling tasks. For instance, the you define a
+service that instructs the orchestrator to keep three instances of a HTTP
+listener running at all times. The orchestrator responds by creating three
+tasks. Each task is a slot that scheduler fills by spawning a container. The
```
Comment (aaronlehmann) [inline docs/swarm/how-swarm-mode-works/services.md]: "a slot that the scheduler fills"

File: docs/swarm/how-swarm-mode-works/services.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+schedules tasks to worker nodes.
+
+![services flow](../images/service-lifecycle.png)
+
+## Replicated and global services
+
+There are two types of service deployments, replicated and global.
+
+For a replicated service, you specify the number of identical tasks you want to
+run.  In example, you decide to deploy an HTTP service with three replicas, each
```
Comment (aaronlehmann) [inline docs/swarm/how-swarm-mode-works/services.md]: Nit: extra space

File: docs/swarm/how-swarm-mode-works/services.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+serving the same content.
+
+A global service is a service that runs one task on every node.  There is no
```
Comment (aaronlehmann) [inline docs/swarm/how-swarm-mode-works/services.md]: Nit: extra space

File: docs/swarm/how-swarm-mode-works/services.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
+<!--[metadata]>
++++
+aliases = [
+"/engine/swarm/how-swarm-mode-works/"
```
Comment (thaJeztah) [inline docs/swarm/how-swarm-mode-works/services.md]: this alias should be removed, as the same alias is on "nodes.md"

**Comment by thaJeztah** [comment]
LGTM, I'll fix https://github.com/docker/docker/pull/24955#discussion_r72402925 in a follow up

File: docs/swarm/how-swarm-mode-works/services.md (hunk 8 above)
Comment (sfsmithcha) [inline docs/swarm/how-swarm-mode-works/services.md]: @thaJeztah , if possible, i think you should leave it on nodes.md. there's no index.md for this, so if someone manually navigates to "/engine/swarm/how-swarm-mode-works/" it directs them to nodes.

Should definitely be removed from services.md

Comment (thaJeztah) [inline docs/swarm/how-swarm-mode-works/services.md]: @sfsmithcha yup, that's the plan, and already fixed, see: https://github.com/docker/docker/issues/25118

**Pull request merged**